packages/ui/src/InteractiveList.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { FocusEvent, useCallback, useMemo, useState } from 'react'
+import { TextField, TextFieldExtraProps } from './TextField'
+import { IconButton } from './IconButton'
+import { Plus, X } from 'react-feather'
+import styles from './InteractiveList.module.scss'
+import { useUID } from 'react-uid'
+import { FieldArrayRenderProps, FieldValidator } from 'formik'
+
+export type InteractiveListExtraProps = {
+  customAddElement?: React.FC<{ setValue: (value: string) => void }>
+} & Pick<TextFieldExtraProps<'text'>, 'label' | 'helperText' | 'inputIcon' | 'type'>
+
+export type InteractiveListCoreProps<V> = {
+  name: string
+  value: string[]
+  onBlur?: (e: FocusEvent<HTMLTextAreaElement>) => void
+  error?: string
+  validate?: FieldValidator
+  arrayHelpers: FieldArrayRenderProps
+  onError: (str: string) => void
+} & InteractiveListExtraProps
+
+export type InteractiveListProps<V> = InteractiveListExtraProps & InteractiveListCoreProps<V>
+
+const InteractiveList = <V extends object>({
+  label,
+  error,
+  name,
+  customAddElement: CustomAddElement,
+  value,
+  helperText,
+  inputIcon,
+  onError,
+  type,
+  validate,
+  arrayHelpers,
+}: InteractiveListProps<V>) => {
+  const id = useUID()
+  const [inputValue, setValue] = useState<string>('')
+
+  const normalizedValue = useMemo(() => inputValue.trim(), [inputValue])
+
+  const addValueCallback = useCallback(async () => {
+    if (normalizedValue.length === 0) {
+      onError('You need to provide a non empty value')
+    } else if (value.includes(normalizedValue)) {
+      onError('You need to provide a unique value')
+    } else {
+      if (validate) {
+        const error = await validate(normalizedValue)
+        if (error) {
+          onError(error)
+        } else {
+          arrayHelpers.push(normalizedValue)
+          setValue('')
+        }
+      } else {
+        arrayHelpers.push(normalizedValue)
+        setValue('')
+      }
+    }
+  }, [value, normalizedValue])
+
+  return (
+    <>
+      <div className={styles.inputRow}>
+        <TextField
+          id={id}
+          type={type}
+          helperText={helperText}
+          inputIcon={inputIcon}
+          error={typeof error === 'string' ? error : undefined}
+          label={label}
+          name={name}
+          value={inputValue}
+          onChange={({ target: { value } }) => {
+            setValue(value)
+          }}
+          onKeyPress={async (e) => {
+            if (e.key === 'Enter') {
+              e.preventDefault()
+              await addValueCallback()
+            }
+          }}
+        />
+        <IconButton
+          kind="transparent"
+          ariaLabel="Add Icon"
+          icon={Plus}
+          className={styles.addIcon}
+          size="normal"
+          onClick={async () => {
+            await addValueCallback()
+          }}
+        />
+      </div>
+      {CustomAddElement && <CustomAddElement setValue={setValue} />}
```

Comment:
Do you think we could limit it to a button for now? We could replicate `actions` in `ModalProps`, but make it a single optional action, not an array. for consistency. WDYT?

Reply:
Sure no problem.

Reply:
So I've had a look and passing a component is probably the best solution? The point is we pass this `setValue` pointer that can change the internal state of our component. I'm not sure how this would be doable with approach like we see in ModalProps.

Reply:
What if we exposed `setValue` via `useImperativeHanlde` and let any arbitrary JSX here? Then we could have more freedom in terms of how we can control the state of InteractiveList and at the same time we could use the `setValue` from the exposed handle in the JSX that we wrap. WDYT?